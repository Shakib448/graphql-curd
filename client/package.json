{
  "name": "client",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "test": "jest"
  },
  "dependencies": {
    "@apollo/client": "^3.4.17",
    "@types/enzyme": "^3.10.10",
    "graphql": "^16.0.1",
    "next": "12.0.3",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "waait": "^1.0.5"
  },
  "jest": {
    "setupFilesAfterEnv": [
      "./src/setupTests.ts"
    ]
  },
  "devDependencies": {
    "@testing-library/dom": "^8.11.1",
    "@testing-library/jest-dom": "^5.15.0",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.0.2",
    "@types/node": "^16.11.7",
    "@types/react": "^17.0.35",
    "@types/react-dom": "^17.0.11",
    "@types/react-test-renderer": "^17.0.1",
    "@typescript-eslint/eslint-plugin": "^5.3.1",
    "babel-jest": "^27.3.1",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.6",
    "eslint": "7",
    "eslint-config-next": "12.0.3",
    "jest": "^27.3.1",
    "jest-enzyme": "^7.1.2",
    "jsdom": "^18.1.0",
    "jsdom-global": "^3.0.2",
    "react-test-renderer": "^17.0.2",
    "ts-node": "^10.4.0",
    "typescript": "^4.4.4"
  }
}
