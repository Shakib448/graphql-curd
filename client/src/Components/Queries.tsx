import { useQuery } from "@apollo/client";
import { GET_ALL_USERS } from "../Graphql/Queries";

const Queries = ({ data, dataLoading }) => {
  if (dataLoading) return "Loading...";
  return (
    <div>
      {data.map((item: any) => (
        <div key={item.id}>
          <h1>name : {item.name}</h1>
          <h1>username : {item.username}</h1>
        </div>
      ))}
    </div>
  );
};

export default Queries;
