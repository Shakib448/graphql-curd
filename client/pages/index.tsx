import React, { useState } from "react";
import Head from "next/head";
import { useMutation } from "@apollo/client";
import Link from "next/link";

import { CREATE_USER } from "../src/Graphql/Mutation";

const Home = () => {
  const [name, setName] = useState("");
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [createUser, { error, loading }] = useMutation(CREATE_USER);

  if (loading) return "Submitting...";
  if (error) return `Submission error! ${error.message}`;

  return (
    <div>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main>
        <div className="createUser">
          <input
            type="text"
            placeholder="name"
            onChange={(e) => setName(e.target.value)}
          />
          <input
            type="text"
            placeholder="username"
            onChange={(e) => setUsername(e.target.value)}
          />
          <input
            type="text"
            placeholder="password"
            onChange={(e) => setPassword(e.target.value)}
          />
          <button
            onClick={() =>
              createUser({
                variables: {
                  name: name,
                  username: username,
                  password: password,
                },
              })
            }
          >
            submit
          </button>
          <Link href="/newUser">GO</Link>
        </div>
      </main>
    </div>
  );
};

export default Home;
