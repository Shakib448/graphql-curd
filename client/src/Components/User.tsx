import React from "react";

import { useMutation, useQuery } from "@apollo/client";
import { DELETE_USER } from "../Graphql/Mutation";
import { GET_ALL_USERS } from "../Graphql/Queries";

const User = (): JSX.Element => {
  const [deleteUser, { error: mutationError }] = useMutation(DELETE_USER);
  const { data, loading, error: queryError } = useQuery(GET_ALL_USERS);

  if (loading) {
    return <h1>Loading...</h1>;
  }
  if (queryError) return <h1>Error!</h1>;
  if (mutationError) return <h1>An error has occurred!</h1>;
  return (
    <div>
      {data &&
        data.getAllUsers.map(
          (item: { id: string; name: string; username: string }) => (
            <div key={item.id}>
              <h1>name : {item.name} </h1>
              <h1>username : {item.username}</h1>
              <button
                onClick={() => deleteUser({ variables: { id: item.id } })}
              >
                Delete
              </button>
            </div>
          )
        )}
    </div>
  );
};

export default User;
