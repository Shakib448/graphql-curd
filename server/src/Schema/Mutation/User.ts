import { GraphQLString } from "graphql";
import { UserType } from "../TypeDefs/User";

export const CREATE_USER = {
  type: UserType,
  args: {
    name: { type: GraphQLString },
    username: { type: GraphQLString },
    password: { type: GraphQLString },
  },
  resolve(parent: any, args: any) {
    const { name, username, password } = args;
    return args;
  },
};
