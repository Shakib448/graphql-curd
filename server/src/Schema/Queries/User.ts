import { GraphQLList } from "graphql";
import { UserType } from "../TypeDefs/User";

export const GET_ALL_USER = {
  type: new GraphQLList(UserType),
  resolve() {
    return [];
  },
};
