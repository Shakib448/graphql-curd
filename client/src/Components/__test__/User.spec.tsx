/* eslint-disable testing-library/no-wait-for-snapshot */
/* eslint-disable testing-library/await-async-query */
import { render, screen, waitFor } from "@testing-library/react";
import User from "../User";
import { MockedProvider } from "@apollo/client/testing";
import { GET_ALL_USERS } from "../../Graphql/Queries";
import { GraphQLError } from "graphql";
import { DELETE_USER } from "../../Graphql/Mutation";
import userEvent from "@testing-library/user-event";

describe("Testing with renders", () => {
  it("renders with loading", () => {
    const mocks: any[] = [];
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <User />
      </MockedProvider>
    );
    const loadingText = screen.getByText(/Loading.../i);
    expect(loadingText).toBeInTheDocument();
  });

  it("Should render by length users data and also data", async () => {
    const mocks = {
      request: {
        query: GET_ALL_USERS,
      },
      result: {
        data: {
          getAllUsers: [
            { id: 1, name: "shakib", username: "muktadir" },
            { id: 2, name: "shakib1", username: "muktadir1" },
          ],
        },
      },
    };

    render(
      <MockedProvider mocks={[mocks]} addTypename={false}>
        <User />
      </MockedProvider>
    );

    await waitFor(async () => {
      const text = screen.getAllByRole("heading", {
        name: /username/i,
      });

      expect(text).toHaveLength(2);
      expect(Array.isArray(text)).toBeTruthy();

      expect(text).toMatchSnapshot();
    });
  });

  it("should show error UI", async () => {
    const mocks = {
      request: {
        query: GET_ALL_USERS,
      },
      result: {
        errors: [new GraphQLError("Error!")],
      },
    };

    render(
      <MockedProvider mocks={[mocks]} addTypename={false}>
        <User />
      </MockedProvider>
    );
    await waitFor(async () => {
      const errorText = screen.getByText(/error!/i);

      expect(errorText).toBeInTheDocument();
    });
  });

  it("should delete with success message", async () => {
    const mocks = {
      request: {
        query: GET_ALL_USERS,
      },
      result: {
        data: {
          getAllUsers: [{ id: "1", name: "shakib", username: "muktadir" }],
        },
      },
    };
    const deleteUser = { message: "Successfully Delete" };
    const deleteMockUser: any = {
      request: {
        query: DELETE_USER,
        variables: { id: "1" },
      },
      result: {
        data: {
          deleteUser,
        },
      },
    };

    render(
      <MockedProvider mocks={[mocks, deleteMockUser]} addTypename={false}>
        <User />
      </MockedProvider>
    );

    await waitFor(async () => {
      const button = screen.getByRole("button", { name: /delete/i });
      userEvent.click(button);
    });

    const successDeleteMessage = screen.getByText(/Successfully delete!/i);
    expect(successDeleteMessage).toBeInTheDocument();
    expect(successDeleteMessage).toMatchSnapshot();
  });
});
