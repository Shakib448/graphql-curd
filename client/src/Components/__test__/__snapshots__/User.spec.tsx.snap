// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing with renders Should render by length users data and also data 1`] = `
Array [
  <h1>
    username : 
    muktadir
  </h1>,
  <h1>
    username : 
    muktadir1
  </h1>,
]
`;

exports[`Testing with renders should delete with success message 1`] = `
<p>
  Successfully delete!
</p>
`;
